ntp-daemon/src/main.rs

```diff
@@ -11,17 +16,88 @@ async fn main() -> Result<(), Box<dyn Error>> {
 
     let mut config = Config::from_args(args.config, args.peers).await?;
 
-    // Sentry has a guard we need to keep alive,
-    // so store it. The compiler will optimize
-    // this away when not using sentry.
+    // Sentry has a guard we need to keep alive, so store it.
+    // The compiler will optimize this away when not using sentry.
     let _guard = finish_tracing_init(&mut config, has_log_override)?;
 
-    ntp_daemon::spawn(&config.system, &config.peers).await
+    let peers_reader = Arc::new(tokio::sync::RwLock::new(Peers::default()));
+    let peers_writer = peers_reader.clone();
+
+    let socket_directory = config.sockets;
+
+    let main_loop_handle = tokio::spawn(async move {
+        ntp_daemon::spawn(&config.system, &config.peers, peers_writer).await
+    });
+
+    let peer_state_handle =
+        tokio::spawn(peer_state_observer(socket_directory, peers_reader.clone()));
+
+    // exit if any of the tasks has completed
+    tokio::select! {
+        done = main_loop_handle => Ok(done??),
+        done = peer_state_handle => Ok(done??),
+    }
+}
+
+async fn peer_state_observer(
+    socket_directory: PathBuf,
+    peers_reader: Arc<tokio::sync::RwLock<Peers>>,
+) -> std::io::Result<()> {
+    let socket_directory = &socket_directory;
+
+    // create the path if it does not exist
+    std::fs::create_dir_all(socket_directory)?;
+
+    let observe_socket_path = socket_directory.join("observe");
+    let peers_listener = UnixListener::bind(&observe_socket_path)?;
+
+    // this binary needs to run as root to be able to adjust the system clock.
+    // by default, the socket inherits root permissions, but the client should not need
+    // elevated permissions to read from the socket. So we explicitly set the permissions
+    let permissions: std::fs::Permissions = PermissionsExt::from_mode(0o777);
+    std::fs::set_permissions(&observe_socket_path, permissions)?;
+
+    let mut observed = Vec::with_capacity(8);
+
+    loop {
+        let (stream, _addr) = peers_listener.accept().await?;
+
+        let buffer = {
+            let state = peers_reader.read().await;
+
+            observed.clear();
+            observed.extend(state.observe());
+
+            serde_json::to_vec(&observed).unwrap()
+        };
+
+        write_to_unix_socket(stream, &buffer).await?;
+    }
+}
+
+async fn write_to_unix_socket(mut stream: UnixStream, bytes: &[u8]) -> std::io::Result<()> {
+    loop {
+        // Wait for the socket to be writable
+        stream.writable().await?;
+
+        // Try to write data, this may still fail with `WouldBlock`
+        // if the readiness event is a false positive.
+        match stream.try_write(bytes) {
+            Ok(_) => {
+                return stream.shutdown().await;
+            }
+            Err(ref e) if e.kind() == std::io::ErrorKind::WouldBlock => {
+                continue;
+            }
+            Err(e) => {
+                return Err(e);
+            }
+        }
+    }
 }
```

Comment:
Why do we need this utility function, isn't write() in the AsyncWriteExt trait sufficient for our purposes?

Reply:
I can't figure this out. But the above is what the documentation on `UnixStream` suggests. As far as I can see, the `write()` function is still susceptible to the readiness event being a false positive.

Reply:
this case is handled deep inside the stack, so using the `AsyncWriteExt` methods gives the same functionality.